src/Shared/Microsoft.AspNetCore.Razor.Serialization.Json/ObjectReaders_TagHelpers.cs

```diff
@@ -0,0 +1,283 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.Language;
+#if JSONSERIALIZATION_ENABLETAGHELPERCACHE
+using Microsoft.AspNetCore.Razor.Utilities;
+#endif
+using static Microsoft.AspNetCore.Razor.Language.RequiredAttributeDescriptor;
+
+namespace Microsoft.AspNetCore.Razor.Serialization.Json;
+
+internal static partial class ObjectReaders
+{
+    private static readonly StringCache s_stringCache = new();
+
+    [return: NotNullIfNotNull(nameof(str))]
+    private static string? Cached(string? str)
+    {
+        if (str is null)
+        {
+            return null;
+        }
+
+        // Some of the strings used in TagHelperDescriptors are interned by other processes,
+        // so we should avoid duplicating those.
+        var interned = string.IsInterned(str);
+        if (interned != null)
+        {
+            return interned;
+        }
+
+        // We cache all our stings here to prevent them from ballooning memory in our Descriptors.
+        return s_stringCache.GetOrAddValue(str);
```

Comment:
Can we use `String.Intern` instead?

Reply:
Looking more closely, it seems that there's a lot of effort to avoid holding onto string instances. This just checks to see if a string is already interned, and if it is, returns the interned string. If `String.Intern` were used here, it would force every unique string to live beyond its use, which would seem to work against the intent.